src/Agent/NewRelic/Profiler/Common/AssemblyVersion.h

```diff
@@ -0,0 +1,192 @@
+            Revision = metadata.usRevisionNumber;
+        }
+
+        /// <summary>
+        /// Copy constructor
+        /// </summary>
+        /// <param name="other"></param>
+        AssemblyVersion(const AssemblyVersion& other)
+        {
+            Major = other.Major;
+            Minor = other.Minor;
+            Build = other.Build;
+            Revision = other.Revision;
+        }
+
+        /// <summary>
+        /// Helper function to create an AssemblyVersion from a version string. If the
+        /// string is invalid it returns null
+        /// </summary>
+        /// <param name="version">Version string. Only digits and periods are allowed.
+        /// Partial versions (e.g. "2.3" or even just "1") are fine, but prerelease,
+        /// beta, etc. identifiers are not valid</param>
+        /// <returns>An AssemblyVersion or null</returns>
+        static AssemblyVersion* Create(xstring_t version)
+        {
+            if (version.size() == 0)
+            {
+                return nullptr;
+            }
+            auto identifiers = Configuration::Strings::Split(version, '.');
+            int major = 0;
+            int minor = 0;
+            int build = 0;
+            int revision = 0;
+
+            // We could just take the first four numbers but safer to treat as an error
+            if (identifiers.size() > 4)
+            {
+                LogWarn(version, L" is not a valid version");
+                return nullptr;
+            }
+
+            try
+            {
+                if (identifiers.size() > 0) major = xstoi(identifiers[0]);
+                if (identifiers.size() > 1) minor = xstoi(identifiers[1]);
+                if (identifiers.size() > 2) build = xstoi(identifiers[2]);
+                if (identifiers.size() > 3) revision = xstoi(identifiers[3]);
+            }
+            catch (...)
+            {
+                // stoi throws if it can't parse the string
+                LogWarn(version, L" is not a valid version");
+                return nullptr;
+            }
+
+            // Negative numbers aren't valid
+            if ((major < 0) || (minor < 0) || (build < 0) || (revision < 0))
+            {
+                LogWarn(version, L" is not a valid version");
+                return nullptr;
+            }
+
+            // Nor is anything over 65535
+            constexpr int maxValid = (std::numeric_limits<unsigned short>::max)();
+            if ((major > maxValid) || (minor > maxValid) || (build > maxValid) || (revision > maxValid))
+            {
+                LogWarn(version, L" is not a valid version");
+                return nullptr;
+            }
+
+            // Or 0.0.0.0
+            if ((major == 0) && (minor == 0) && (build == 0) && (revision == 0))
+            {
+                return nullptr;
+            }
+
+            return new AssemblyVersion((unsigned short)major, (unsigned short)minor, (unsigned short)build, (unsigned short)revision);
```

Comment:
Should we use the `make_unique` template function here instead of the `new` keyword so that we enforce that all callers of this code will use a `unique_ptr` to manage this instance?

Reply:
I considered that and... unfortunately can't remember why I decided against it. I'll have to think on it a bit.